### PR TITLE
Admin endpoint

### PR DESCRIPTION
**Pull request by stefanorosanelli**
This PR solves #1292 

First minimal implementation of `/admin` endpoint 

Through `/admin/{resource}` and usual POST, GET, PATCH, DELETE methods some resources are managed namely:
  * `/admin/applications` 
  * `/admin/async_jobs`
  * `/admin/ednpoints`
  * `/admin/config`

No relationships are displayed for `applications` and `endpoints` - no available for `config` and `async_jobs` 

A new trait `JsonApiAdminTrait` has been introduced to override default JSON API links and relationships display.

Some checks must be introduced at model/table level to avoid dangerous operations like overriding API KEY in use or writing malformed configurations. This will be addressed in a next issue.

### CODE REVIEW

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                [
+                    'content' => 'Another Configuration',
+                 ],
+            ],
+        ];
+    }
+
+    /**
+     * Test CRUD on admin resources
+     *
```
Comment (didoda) [inline plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php:86]: missing `@param $type`

**Review by didoda — APPROVED** [review]

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php (hunk 1 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php:86]: :+1:      
done!

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+     * @var array
+     */
+    public $fixtures = [
+        'plugin.BEdita/Core.config'
+    ];
+
+    /**
+     * Data provider for `testResource`
+     */
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php]: Missing `@return`

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php (hunk 1 above)
Comment (batopa) [reply plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php:86]: Missing `@return` too

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        $this->assertContentType('application/vnd.api+json');
+
+        $locationHeader = $this->_response->getHeaderLine('Location');
+        $this->assertNotEmpty($locationHeader);
+        $resourceId = substr($locationHeader, strrpos($locationHeader, '/') + 1);
+
+        // VIEW
+        $this->configRequestHeaders();
+        $this->get("$endpoint/$resourceId");
+        $result = json_decode((string)$this->_response->getBody(), true);
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php]: unused `$result` variable. Missing a test on it?

File: plugins/BEdita/API/tests/TestCase/Controller/AdminControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            'type' => 'applications',
+            'attributes' => [
+                'name' => 'another-app',
+                'description' => 'Another app',
+            ],
+        ];
+
+        $this->configRequestHeaders('POST', $this->getUserAuthHeader());
+        $this->post('/admin/applications', json_encode(compact('data')));
+        $result = json_decode((string)$this->_response->getBody(), true);
```
Comment (batopa) [inline plugins/BEdita/API/tests/TestCase/Controller/AdminControllerTest.php]: `$result` not used

File: plugins/BEdita/Core/tests/TestCase/Model/Entity/JsonApiAdminTraitTest.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+
+namespace BEdita\Core\Test\TestCase\Model\Entity;
+
+use BEdita\Core\Utility\JsonApiSerializable;
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Entity/JsonApiAdminTraitTest.php]: Useless `use` statement for `JsonApiSerializable` and `Hash`

**Review by batopa — CHANGES_REQUESTED** [review]

File: plugins/BEdita/Core/tests/TestCase/Model/Entity/JsonApiAdminTraitTest.php (hunk 5 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/tests/TestCase/Model/Entity/JsonApiAdminTraitTest.php]: done

File: plugins/BEdita/API/tests/TestCase/Controller/AdminControllerTest.php (hunk 4 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/TestCase/Controller/AdminControllerTest.php]: done

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php (hunk 3 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php]: removed for now

a detailed view test would be preferable, maybe not in IntegrationTest?

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php (hunk 2 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php]: added

File: plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php (hunk 1 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/IntegrationTest/AdminResourcesTest.php:86]: added

**Comment by fquffio** [comment]
I like the approach to have a single controller manage multiple resources, however I don't really like having a single `AdminController`. `admin/` prefix sounds like something that might become really complex and "wide".

Rather than a single `AdminController` I'd have a `Controller/Admin` namespace with a `ResourcesController` in it. Routes would then be connected using a [prefix](https://book.cakephp.org/3.0/en/development/routing.html#prefix-routing).

This is not blocking: we could switch to this approach in the future if and when we'll actually need this.

**Comment by stefanorosanelli** [comment]
Agreed, an `/admin` prefix and a `Controller\Admin` namespace are both good ideas.

We could to it, among with other necessary improvements, in one or more new issues.

**Review by batopa — APPROVED** [review]

**Pull request merged**